packages/components/src/navigator/navigator-screen/component.tsx

```diff
@@ -65,12 +72,40 @@ function NavigatorScreen( props: Props, forwardedRef: Ref< any > ) {
 		[ className ]
 	);
 
-	// This flag is used to only apply the focus on mount when the actual path changes.
-	// It avoids the focus to happen on the first render.
-	const [ hasPathChanged, setHasPathChanged ] = useState( false );
+	// Focus restoration
+	const [ isFirstRender, setIsFirstRender ] = useState( true );
 	useEffect( () => {
-		setHasPathChanged( true );
-	}, [ path ] );
+		// Only attempt to restore focus:
+		// - after the first render (to avoid moving focus on page load)
+		// - when the screen becomes visible
+		// - the wrapper ref has been assigned
+		if ( isFirstRender || ! isMatch || ! wrapperRef.current ) {
+			setIsFirstRender( false );
```

Comment:
Feels like setIsFirstRender should always be called here even if isMatch is true otherwise its value doesn't reflect its intent and may cause bugs?

Reply:
I don't really understand why we need ` || ! isMatch || ! wrapperRef.current ` in this condition in that case since this boolean will be false anyway?

Reply:
I tried a slightly different angle and pushed some changes in https://github.com/WordPress/gutenberg/pull/37223/commits/14952338a491931afe103f79df2d5688b084c24c — hope it makes the logic clear, let me know what you think!

Reply:
What ever you're most confident with, it was just a small implementation thing that I was wondering about :) 
The new approach look nicer but wonder about whether someone could actually set `isInitial` using `push` creating a bad behavior.

Reply:
I also think that the new implementation is more clear.

I've also made sure that `isInitial` can't be overridden externally in https://github.com/WordPress/gutenberg/pull/37223/commits/11a72bbd2dc5be7fe73fb060c643486ede461224